text/0000-fabric-gateway.md

```diff
@@ -0,0 +1,207 @@
+  - The `Prepare` service is invoked on the Gateway, passing the SignedProposal message
+    - The Gateway will determine the endorsement plan for the requested chaincode and forward to the appropriate peers for endorsement. It will return to the client a `PreparedTransaction` message which contains a `Envelope` message as defined in `fabric-protos/common/common.proto`.  It will also contain other information, such as the return value of the chaincode function and the transaction ID so that the client doesn't necessarily need to unpack the `Envelope` payload.
+- __Commit__
+  - The client signs the hash of the Envelope payload using the user's private key and sets the signature field in the `Envelope` in the `PreparedTransaction`.
+  - The `Commit` service is invoked on the Gateway, passing the signed `PreparedTransaction` message.  A stream is opened to return multiple return values.
+    - The Gateway will register transaction event listeners for the given channel/txId.
+    - It will then broadcast the `Envelope` to the ordering service.
+    - The success/error response is passed back to the client in the stream
+    - The Gateway awaits suffient tx commit events before returning and closing the stream, indicating to the client that transaction has been committed.
+
+Evaluating a transaction is a simple process of invoking the `Evaluate` service passing a SignedProposal message.  The Gateway passes the request to a peer of it's choosing according to a defined policy (probably same org, highest block count) and returns the chaincode function return value to the client.
+
+## Launching the Gateway
+
+To run the gateway server, the following parameters need to be supplied:
+- The url of at least one peer in the org.  Once connected, the discovery service will be invoked to find other peers.
+- The MSPID associated with the gateway.
+- The signing identity of the gateway (cert and key), e.g. location of PEM files on disk.
+- The TLS certificate so that the gateway can connect to other components in the organization
+The following is an example command line invocation using a prototype:
+- `gateway -h peer0.org1.example.com -p 7051 -m Org1MSP -id ../../fabric-samples/fabcar/javascript/wallet/gateway.id -tlscert ../../fabric-samples/test-network/organizations/peerOrganizations/org1.example.com/tlsca/tlsca.org1.example.com-cert.pem`
+
+Alternatively, the Gateway server will be designed to be embeddable within a Peer, in which case the all of the above parameters will be taken from the peer configuration.
+
+### Discovery
+
+On startup, the gateway will connect to its primary peer(s), as specified in the command line option (or its host, if embedded in a peer).  The discovery service will be invoked to build and maintain a cache of the network topology per channel.  This will be used to identify endorsers based on the chaincode endorsement policy.
```

Comment:
Right now, the design delegates signing the proposal and the transaction to the client, however here you are saying that the gateway requires an identity and a corresponding private key.

The discovery service authenticates clients if they are in the channel writers policy. This means, that if you can query discovery, you can also submit transactions. 

I am wondering if it would make sense, to delegate the signature on discovery queries to the client.

If we can do that, then you could deploy a gateway without any enrollment certificate at all, and this would potentially secure the whole deal much more, because the said gateway cannot craft transactions on its own, so, if its file system is somehow compromised, no private key corresponding to an enrollment identity would be stolen.

Another reason why this fits well with discovery, is that discovery queries do not have nonces or timestamps, and are meant to be re-used: If you submit a query Q signed with signature S twice, the peer would only perform verification of signature for the first time, and the second time it would lookup in the cache. 

Therefore I think you should consider further delegating the signatures on discovery queries to the client.

Reply:
This is an interesting suggestion. I'm not sure how workable it is. If I describe the issues as I see them then perhaps you can see ways to make this work.

The Gateway is intended stateless and scalable, so a single Gateway instance is used by multiple clients, and a client may interact with different Gateway instances on different requests, which may or may not require discovery. So it seems like we might have to have a chatty interaction where a client generates and signs a discovery query on every request. Or at least checks whether it needs to sign a discovery query, which is equally (if not more) chatty.

Gateway will need to listen to events from peers. I think this will also require signing?

I think the Gateway might at some point need to be able to evaluate transactions, such as the `getMetadata` transactions in chaincode to obtain the chaincode metadata. I am thinking of this as a way of obtaining information needed for endorsement in more complex scenarios, such as identifying the collections used by specific transaction functions, or chaincode-to-chaincode calls made by specific transaction functions.

We've got some mechanisms we could use to improve security of server-side credentials, such as injecting Kubernetes secrets and/or using HSM.

Any signing requirements for the Gateway implementation we expose (or delegate) down to the client introduces tighter coupling between client and the specific point-in-time implementation of the Gateway. Ideally we want to minimise the interface between client and Gateway as much as possible, and to expose as little of the Gateway logic to the client as possible. Requiring client-side signature for things like discovery would trade off against that goal.

The Gateway is running server-side in the same client environment as your peers and orders. If your server-side is compromised to that degree then I guess all your nodes and their ledgers are similarly compromised. Given that, how much are we buying by removing the need for the Gateway to have its own signing identity?

Reply:
I'm failing to see the issue. Even if a single gateway is used by multiple clients, you can have an in-memory segment for each client, right? 

As for the chatting - Imagine that you're a client and you're sending the gateway some RPC `Foo` and expect to get some `X` in return. 
It (the gateway) either sends you back that `X`, or, instead, it sends you back (`Y`, `h`) - something to sign (`Y`), and a handle `h` (a unique string) which you can then invoke the same RPC call `Foo` with that handle `h` as a parameter to retrieve `X` which should be ready by then. 

> Gateway will need to listen to events from peers. I think this will also require signing?

Yes but when do you listen to events? when you submit a transaction beforehand, no? Since when do clients simply listen to events just for the sake of events? They listen for events to know the status of their transaction.



> I think the Gateway might at some point need to be able to evaluate transactions, such as the getMetadata transactions in chaincode to obtain the chaincode metadata. I am thinking of this as a way of obtaining information needed for endorsement in more complex scenarios, such as identifying the collections used by specific transaction functions, or chaincode-to-chaincode calls made by specific transaction functions.

Are you saying that the gateway is expected to be able to invoke chaincode on its own behalf ? How would it know the parameters of the chaincode? How would the developer of the chaincode know what parameters the gateway expects? It seems to me that you can just instead add an API to the peer to obtain metadata, if that's all you want. 

> Any signing requirements for the Gateway implementation we expose (or delegate) down to the client introduces tighter coupling between client and the specific point-in-time implementation of the Gateway. Ideally we want to minimise the interface between client and Gateway as much as possible, and to expose as little of the Gateway logic to the client as possible. Requiring client-side signature for things like discovery would trade off against that goal.

Understood.


> The Gateway is running server-side in the same cloud environment as your peers and orders. If your server-side is compromised to that degree then I guess all your nodes and their ledgers are similarly compromised. Given that, how much are we buying by removing the need for the Gateway to have its own signing identity?

The key might leak in other ways, like backups, compromising only a single node (host) of the cloud environment but not be able to compromise all of them.

Additionally, there are other vectors to extract a private key than to take over the host. For example, if you give the server to sign a lot of things and the signing is **not** constant time, then by collecting enough samples, you can extract some bits of the private key because you know (statistically) how much time it takes to do certain math operations, so you divide and conquer the key space until you end up recovering the key completely. 

Imagine that I somehow made your gateway to send me lots and lots of discovery requests, because I either have compromised a public TLS CA (it happened in the past) and issued a certificate with the same SAN to myself or from some reason, a TLS private-public key pair was generated with bad randomness and I brute forced the certificate and found the private key. 

Having said that, I guess that any problem that is possible with the gateway, you might already have with the SDK.

Reply:
> Yes but when do you listen to events? when you submit a transaction beforehand, no? Since when do clients simply listen to events just for the sake of events? They listen for events to know the status of their transaction.
Clients absolutely can (and do) listen to events without submitting transactions, and seeing whether a submitted transaction has committed successfully is absolutely not the only reason clients listen to events.

Reply:
> Are you saying that the gateway is expected to be able to invoke chaincode on its own behalf ? How would it know the parameters of the chaincode? How would the developer of the chaincode know what parameters the gateway expects? It seems to me that you can just instead add an API to the peer to obtain metadata, if that's all you want.

https://hyperledger.github.io/fabric-chaincode-node/release-2.2/api/tutorial-annotated-contract-metadata.html

I don't think the peer holds the information about a given transaction function implementation that I described.

Reply:
> As for the chatting - Imagine that you're a client and you're sending the gateway some RPC Foo and expect to get some X in return.
It (the gateway) either sends you back that X, or, instead, it sends you back (Y, h) - something to sign (Y), and a handle h (a unique string) which you can then invoke the same RPC call Foo with the signed Y and that handle h as a parameter to retrieve X which should be possible to retrieve now that the gateway has the discovery signature.

As you've demonstrated, that is more chatty than the proposed interaction. It sounds like you are proposing adding session state into the stateless Gateway.

Reply:
> Additionally, there are other vectors to extract a private key than to take over the host...

How are these vulnerabilities address in the peer and orderer today?

Reply:
> > Additionally, there are other vectors to extract a private key than to take over the host...
> 
> How are these vulnerabilities addressed in the peer and orderer today?

We always use curve P-256 which is constant time. When I wrote the comment I thought you intended to implement the gateway using several languages.